service/src/prometheus_metrics.rs

```diff
@@ -0,0 +1,161 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+//! Service for prometheus metrics, hosted on a http server.
+
+use crate::{
+	account_funding::EnclaveWalletInfo,
+	error::{Error, ServiceResult},
+};
+use async_trait::async_trait;
+use itp_enclave_metrics::EnclaveMetric;
+use lazy_static::lazy_static;
+use log::*;
+use prometheus::{proto::MetricFamily, register_int_gauge, IntGauge};
+use std::{net::SocketAddr, sync::Arc};
+use warp::{Filter, Rejection, Reply};
+
+lazy_static! {
+	static ref ENCLAVE_ACCOUNT_FREE_BALANCE: IntGauge =
+		register_int_gauge!("enclave_account_free_balance", "Free balance of the enclave account")
+			.unwrap();
+	static ref ENCLAVE_SIDECHAIN_BLOCK_HEIGHT: IntGauge =
+		register_int_gauge!("enclave_sidechain_block_height", "Enclave sidechain block height")
+			.unwrap();
+	static ref ENCLAVE_SIDECHAIN_TOP_POOL_SIZE: IntGauge =
+		register_int_gauge!("enclave_sidechain_top_pool_size", "Enclave sidechain top pool size")
+			.unwrap();
+}
+
+pub async fn start_metrics_server<MetricsHandler>(
+	metrics_handler: Arc<MetricsHandler>,
+	port: u16,
+) -> ServiceResult<()>
+where
+	MetricsHandler: HandleMetrics + Send + Sync + 'static,
+{
+	let metrics_route = warp::path!("metrics").and_then(move || {
+		let handler_clone = metrics_handler.clone();
+		async move { handler_clone.handle_metrics().await }
+	});
```

Comment:
I have problems understanding this code. Why do we need the nested move?

Reply:
This took me 3 hours to figure out - and I still cannot say I understand it fully myself 🙈  

First we create a lambda, into which we move the handler, that is the easy part to understand. But then every request needs its own instance of a request handler, and this is the only way I was able to make that work (it implements `Fn` and not only `FnOnce`, which would be the case if you didn't have the `clone()` like on line 52.

Reply:
Ok, interesting. Well, to me it seems like you did understand the problematic here. 😄